perf(sql): speed up parallel sample by with fill

nwoolmer opened the pull request:
WIP

**Changelist**

Parallel sample by:

- Parallel sample by:
    - now supports fill with multiple values
    - correctly uses all fill values instead of just the first
    - will now generate radix sorts more frequently, which are 2x-10x faster than our tree-based sort
    - will now only sort unfilled data, instead of sorting both filled and real data
 - a mismatch between column and fill value in from-to queries previously caused `UnsupportedOperationException` and a `500` response code. Now it will return a useful message:
    - **Example query:** `select timestamp, count from trades sample by 1d from '2008-12-28' to '2009-01-05' FILL(1.0)`
    - **Error**: `invalid fill value, cannot cast DOUBLE to LONG`


**Benchmarks**

Table of 100m rows.
Schema - (TIMESTAMP, DOUBLE)
Data is over a 1 day period.

**With 100m rows**

Query:

```sql
select timestamp, avg(price) from bench
sample by 1T fill(null)
```

<details><summary>Plan before</summary>

```
Sort
  keys: [timestamp]
    Fill Range
      stride: '1T'
      values: [null]
        Async Group By workers: 8
          keys: [timestamp]
          values: [avg(price)]
          filter: null
            PageFrame
                Row forward scan
                Frame forward scan on: bench
````
</details>

<details><summary>Plan after</summary>

```
Fill Range
  stride: '1T'
  values: [null]
    Radix sort light
      keys: [timestamp]
        Async Group By workers: 8
          keys: [timestamp]
          values: [avg(price)]
          filter: null
            PageFrame
                Row forward scan
                Frame forward scan on: bench
```
</details>


Before: `timeout (60s)`
After: `22.75s`

**With 10m rows:**

Query:
```sql
select timestamp, price from (
select timestamp, avg(price) as price from bench4
sample by 1T fill(null)
) limit -1
```

<details><summary>Before plan:</summary>

```
Limit lo: -1
    Sort
      keys: [timestamp]
        Fill Range
          stride: '1T'
          values: [null]
            Async Group By workers: 8
              keys: [timestamp]
              values: [avg(price)]
              filter: null
                PageFrame
                    Row forward scan
                    Frame forward scan on: bench4
```
</details>

<details><summary>After plan:</summary>

```
Limit lo: -1
    Fill Range
      stride: '1T'
      values: [null]
        Radix sort light
          keys: [timestamp]
            Async Group By workers: 8
              keys: [timestamp]
              values: [avg(price)]
              filter: null
                PageFrame
                    Row forward scan
                    Frame forward scan on: bench4
```
</details>


`Before: 13.4s, 12.51s, 14.4s`
`After: 1.9s, 1.43s, 1.33s`

Queries which do not require the `sort` will be slower, for example:

```sql
select max(price) from (
    select timestamp, avg(price) from bench
    sample by 1T fill(null)
)
```

This is because in the new version, a sort will be generated, and in the old, it would not. However, this was at the cost of the fill not necessarily being correct or filling on the right timestamp.

**Alternative**

If we allowed group by to return a `timestampIndex` in this case, even though the result will not be sorted (i.e `(timestamp, SCAN_DIRECTION_OTHER)`, then we could still use parallel fill in its original form. This requires relaxing the rule that `timestampIndex` means `ASC` (designated) timestamp.

**Notes**

This PR reworks the parallel sample by fill in an effort to improve performance and reliability. Previously, the fill would be generated after group by, and before order by. This meant we had to first exhaust the group by result, recording timestamps. Then we filled the rest of the timestamps. 

Unfortunately, group by does not guarantee a designated timestamp, making it problematic to identify the correct column to fill. Furthermore, this would cause both filled and unfilled data would be subsequently sorted, so sparse datasets were as expensive to sort as dense datasets.

By moving it to after order by, we ensure that we have a designated timestamp to orient the fill. We also generate `radix sort` more frequently, which can be 2-10x faster than our tree-based sort. We also only have to sort the unfilled data, leading to a much quicker an more efficient sort.

nwoolmer committed:
add some basic validation to catch mismatched type error earlier

nwoolmer committed:
fix condition

nwoolmer committed:
fix SqlOptimiserTest

nwoolmer committed:
condition was back to front

glasstiger commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 13 / 13 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java|13|13|100.00%|

amunra commented:
Not really code you've directly touched, just stumbled on it via your tests. The following error message is rather cryptic.

```java
    private void guardAgainstFromToWithKeyedSampleBy(boolean isFromTo) throws SqlException {
        if (isFromTo) {
            throw SqlException.$(0, "FROM-TO intervals are not supported for keyed SAMPLE BY queries");
        }
    }
```

I have no idea what a keyed query is.

amunra reviewed (dismissed):
👍

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java:
```diff
@@ -93,7 +94,34 @@ public FillRangeRecordCursorFactory(
+
+                int fillColumnType = valueFuncs.get(i - passedTimestamp).getType();
+                if (fillColumnType != columnType && !ColumnType.isBuiltInWideningCast(fillColumnType, columnType)) {
+                    throw SqlException.$(0, "invalid fill value, cannot cast " + ColumnType.nameOf(fillColumnType) + " to " + ColumnType.nameOf(columnType));
```
SqlException has a machanism to avoid string concatentation, please use it. cc @amunra

bluestreak01 replied:
same here, error position must not be 0, especially for a fill value cc @amunra

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java:
```diff
@@ -93,7 +94,34 @@ public FillRangeRecordCursorFactory(
+            int passedTimestamp = 0;
+
+            // validate metadata
+            for (int i = 0; i < metadata.getColumnCount(); i++) {
```
please use "standard" loop, e.g. `for (int i = 0, n = metadata.getColumnCount(); i < n ...` cc @amunra

bluestreak01 reviewed (changes requested):
cc @amunra

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java:
```diff
@@ -93,7 +94,34 @@ public FillRangeRecordCursorFactory(
+                throw SqlException.$(0, "not enough fill values");
+            }
+
+            int passedTimestamp = 0;
```
not enough tests for the `passedTimestamp` calculations. Also, please comment what the code on what the logic for `passedTimestamp`

nwoolmer committed:
refactor to pass fill values positions

nwoolmer committed:
comments

nwoolmer committed:
fix condition

nwoolmer committed:
Merge branch 'master' into nw_fill_range_error

bluestreak01 committed:
Merge remote-tracking branch 'origin/master' into nw_fill_range_error

bluestreak01 committed:
merge fallout

bluestreak01 committed:
fixes

glasstiger commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 28 / 29 (96.55%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java|16|17|94.12%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|12|12|100.00%|

bluestreak01 committed:
formatting

glasstiger commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 28 / 29 (96.55%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java|16|17|94.12%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|12|12|100.00%|

bluestreak01 commented on core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:
```diff
@@ -2019,17 +2019,21 @@ private RecordCursorFactory generateFill(QueryModel model, RecordCursorFactory g
         try {
-            if (fillValuesExprs == null) {
-                throw SqlException.$(-1, "fill values were null");
+            IntList fillValuesPos = new IntList(fillValuesExprsSize);
```
we allocate an object even if we don't fill functions (fill count = 0)

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java:
```diff
@@ -94,6 +98,44 @@ public FillRangeRecordCursorFactory(
+            Therefore, once we pass the column, we need to offset back one entry to find
+            the corresponding fill value.
+             */
+            int passedTimestamp = 0;
```
logic with skipping timestamp like this is flawed when columns are not arranged sequentially, .e.g. this fails with array index out of bounds:

```sql
create table y as (select timestamp_sequence(1000000, 100000) ts, rnd_double() t from long_sequence(1000000)) timestamp(ts) partition by day;

select first(t), ts from y
sample by 15m
fill(2.0);

```

nwoolmer committed:
rework parallel sample by fill. now `FillRange` requires a designated timestamp i.e ascending sorted by timestamp. the result is that we generate faster sorts in some cases, and also sort smaller amounts of data (only unfilled)

still an outstanding issue with set operations, where a sort is not being appropriately pushed down, causing the fill to be lost

nwoolmer committed:
fix tests, still todo - fix set operations

nwoolmer committed:
Merge branch 'master' into nw_fill_range_error

nwoolmer committed:
improve error message

nwoolmer committed:
safety

nwoolmer committed:
safety

nwoolmer committed:
still not a fix

nwoolmer committed:
hacky fix, more attention required

nwoolmer committed:
Merge branch 'refs/heads/master' into nw_fill_range_error

# Conflicts:
#	core/src/main/java/io/questdb/griffin/model/QueryModel.java

nwoolmer committed:
Merge branch 'master' into nw_fill_range_error

nwoolmer committed:
fix test
